tests/interfaces/test_batch_autograd_qnode.py

```diff
@@ -0,0 +1,1238 @@
+
+        @qnode(dev, diff_method=diff_method, interface="autograd", mode=mode)
+        def circuit(a, b, c):
+            qml.RY(a * c, wires=0)
+            qml.RZ(b, wires=0)
+            qml.RX(c + c ** 2 + np.sin(a), wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = qml.jacobian(circuit)(a, b, c)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == {0, 2}
+            tape_params = np.array(circuit.qtape.get_parameters())
+            assert np.all(tape_params == [a * c, c + c ** 2 + np.sin(a)])
+
+        assert res.shape == (2,)
+
+    def test_no_trainable_parameters(self, dev_name, diff_method, mode, tol):
+        """Test evaluation and Jacobian if there are no trainable parameters"""
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd", mode=mode)
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        a = np.array(0.1, requires_grad=False)
+        b = np.array(0.2, requires_grad=False)
+
+        res = circuit(a, b)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == set()
+
+        assert res.shape == (2,)
+        assert isinstance(res, np.ndarray)
+
+        assert not qml.jacobian(circuit)(a, b)
+
+        def cost(a, b):
+            return np.sum(circuit(a, b))
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
+            grad = qml.grad(cost)(a, b)
+
+        assert grad == tuple()
+
+    def test_matrix_parameter(self, dev_name, diff_method, mode, tol):
+        """Test that the autograd interface works correctly
+        with a matrix parameter"""
+        U = np.array([[0, 1], [1, 0]], requires_grad=False)
+        a = np.array(0.1, requires_grad=True)
+
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd", mode=mode)
+        def circuit(U, a):
+            qml.QubitUnitary(U, wires=0)
+            qml.RY(a, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit(U, a)
+
+        if diff_method == "finite-diff":
+            assert circuit.qtape.trainable_params == {1}
+
+        res = qml.grad(circuit)(U, a)
+        assert np.allclose(res, np.sin(a), atol=tol, rtol=0)
+
+    def test_gradient_non_differentiable_exception(self, dev_name, diff_method, mode):
+        """Test that an exception is raised if non-differentiable data is
+        differentiated"""
+        dev = qml.device(dev_name, wires=2)
+
+        @qnode(dev, interface="autograd", diff_method=diff_method)
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit, argnum=0)

```

Comment:
So the difference between this test and the previous one is that here we are explicitly telling it to differentiate w.r.t. the argument by setting `argnum=0`? If we didn't set this, it would just say "output seems independent of input"?

Reply:
yep!